### PR TITLE
Signup: Show "Related" topics when searching in site topic step

### PR DESCRIPTION
**Pull request by p-jackson**
Aug-23-2019 18-19-09

#### Changes proposed in this Pull Request

* Adds headings to the `<Suggestions>` component by adding an optional "category" to suggestions, which will be grouped together in the list.
* Adds a "Related" category to the site topic suggestions. Anything vertical that doesn't "string match" query typed by the user will be shown as a "Related" topic.

These changes should have no effect if the vertical search API doesn't return synonyms in the results. **This change should be shipped before D31114-code.**

Some more discussion: paObgF-qR-p2

Feedback welcome on whether "category" is the best term to use for the new `<Suggestions>` feature.

#### Testing instructions

Apply D32511-code.

Follow instructions in D31114-code to add yourself to the 'synonymns' or 'original' test variation.

* Try original variation: vertical picker should work as before, no **Related** header should appear.
* Try synonymns variation:
  * synonyms should only appear in the business segment
  * synonyms that don't match the query value should appear under the **Related** header
  * no related matches appear if the query is less than 3 characters long
  * mouse hover, mouse clicks, and keyboard events should work with the items under the **Related** header.

Fixes Automattic/zelda-private#117

### CODE REVIEW

**Comment by p-jackson** [comment]
Going to do more testing to ensure the railcar tracks are still working (the indexes match the order of the API response), but ready for someone to take a look.

**Comment by ramonjd** [comment]
Looking really nice @p-jackson !!

I noticed that if delete a little more slowly I get related results (unsandboxed)

Sep-03-2019 11-17-18

I haven't looked too closely yet, but would it make your life easier if the backend informed the frontend it was using synonym-based matching?

**Comment by p-jackson** [comment]
@ramonjd Thanks for finding! I was able to sort this out with a string trim so no API change necessary yet 🤞
I feel like adding whether synonym search was active would be a breaking API change (currently the response body is an array) which would make deploying trickier. At may come to needing that though if there are other bugs that need squashed.

**Comment by p-jackson** [comment]
As this change rearranges the results that come from the `/verticals` API I've also added the `ui_algo` and `ui_position` properties to the traintrack events that get fired. Described here: PCYsg-bor-p2

**Comment by ramonjd** [comment]
> I feel like adding whether synonym search was active would be a breaking API change (currently the response body is an array) which would make deploying trickier. At may come to needing that though if there are other bugs that need squashed.

Good call, especially now that I've finally grokked how you're [categorizing results](https://github.com/Automattic/wp-calypso/pull/35724/files#diff-9553ff7100a8a69d6011c3bc6f8dded2R183). ;)

How do you think we should handle the following situation, in which the server escapes certain chars (as far as I can tell it's only `<`)?

Screen Shot 2019-09-04 at 9 23 12 am

I could try to strip them out in the backend.

Also, I'm not sure if it's just me but D31114-code always seems to return synonym matches even though there's a 50/50 split. If I get time today/tomorrow I'll also look at that.

All in all this PR is looking sweet: it tests well and, as far as I'm concerned, is ready to land once we sort out the backend issues.

**Comment by ramonjd** [comment]
> I could try to strip them out in the backend.

I tried this, but it doesn't have much affect as it looks like we're adding the input value to the related category.

Screen Shot 2019-09-04 at 12 24 45 pm

Do you think there's a way to strip out/ignore such chars in the results on the frontend? Or maybe if there's only one result, or the result has `is_user_input: true`, we'd not show the related header?

**Comment by p-jackson** [comment]
😭 I wish `sanitize_text_field` didn't escape only some characters, it'd be a lot easier to display correctly on the client if it escaped everything†

&lt;/rant>

> or the result has `is_user_input: true`, we'd not show the related header?

I think that's the fix! If it's `is_user_input_vertical` then it should never be a *related* topic, it's the main thing they're looking for.

> I feel like adding whether synonym search was active would be a breaking API change (currently the response body is an array) which would make deploying trickier.

Actually I've realised it's more than just that, there's a note in D31114-code that we need to make sure we don't change things on mobile.

---
† it probably just needs to escape `&` as well and you'd be able to render as html in React.

**Comment by dwolfe** [comment]
This looks good to me - the "Related" heading reuses the styling for the "Popular" suggestions we show in the empty state, right? 👍 from me.

**Comment by p-jackson** [comment]
@dwolfe yep they're the same.

The popular and related sections should be exactly the same except for some keyboard navigation and focus differences that I'd like to fix as a follow up to this PR and #35765.

File: client/state/data-layer/wpcom/signup/verticals/index.js [hunk 1]
```diff
@@ -31,6 +31,7 @@ export const requestVerticals = action =>
 				...( action.siteTypeId && { site_type: action.siteTypeId } ),
 				limit: action.limit,
 				include_preview: true,
+				synonyms: true,
```
Comment (ramonjd) [inline client/state/data-layer/wpcom/signup/verticals/index.js]: Following the suggested update to D32511-code

```suggestion
				allow_synonyms: true,
```

File: client/state/data-layer/wpcom/signup/verticals/test/index.js [hunk 2]
```diff
@@ -27,6 +27,7 @@ describe( 'data-layer/wpcom/signup/verticals', () => {
 						site_type: mockAction.siteTypeId,
 						limit: mockAction.limit,
 						include_preview: true,
+						synonyms: true,
```
Comment (ramonjd) [inline client/state/data-layer/wpcom/signup/verticals/test/index.js]: Following the suggested update to D32511-code

```suggestion
				allow_synonyms: true,
```

**Comment by ramonjd** [comment]
@p-jackson 

This is looking 👌 

Screen Shot 2019-09-23 at 2 22 51 pm

Also handles the `<` well. Thanks!

Screen Shot 2019-09-23 at 2 22 41 pm

Once we get D32511-code in do you think we can 🚀 this one?

**Comment by p-jackson** [comment]
@ramonjd I think this is gtg pending a ✅. I'm taking @dwolfe's comment as a design sign-off.

Because of our carefully crafted changes this can either ship before or after  D32511-code :)

**Pull request merged**